<launch>

  <!-- function -->
  <arg name="run_record" default="true" />
  <arg name="run_2d_lidar" default="true" />
  <arg name="run_rviz" default="true" />
  <arg name="run_rscamera" default="true" />

  <node if="$(arg run_rviz)"  name="rviz" pkg="rviz" type="rviz" args="-d $(find lidar)/rviz/record.rviz" required="true" />

  <!-- <arg name="bag_storage_path" default="$(find lidar)/bag/" /> -->
  <!-- <node if="$(arg run_bag)" pkg="rosbag" type="record" name="rosbag_record" args="-o $(arg bag_storage_path)/lidar_2d.bag   /point_cloud /scan" output="screen" /> -->
  <node if="$(arg run_record)"  name="bag_recorder_service" pkg="rtab" type="bag_recorder_service.py" output="screen">
      <!-- <param name="topics_to_record" value="all" /> -->
      <param name="topics_to_record" value="/point_cloud /scan /camera/color/image_raw /odom_robot" />
      <param name="bag_file_directory" value="$(find lidar)/bag" />
  </node>



  <include file="$(find realsense2_camera)/launch/rs_camera.launch" if="$(arg run_rscamera)" >
      <arg name="camera" value="camera" />
      <arg name="serial_no" value="" /> 
      <arg name="usb_port_id" value="" /> 
      <arg name="device_type" value="" />  
      <arg name="color_width" value="640" />
      <arg name="color_height" value="480" /> 
      <arg name="depth_width" value="640" /> 
      <arg name="depth_height" value="480" />
      <arg name="infra_width" value="640" /> 
      <arg name="infra_height" value="480" /> 
      <arg name="color_fps" value="15" /> 
      <arg name="depth_fps" value="15" /> 
      <arg name="infra_fps" value="15" />
      <arg name="align_depth" value="true" />
      <arg name="output" value="log"/>

  </include>

  <node if="$(arg run_2d_lidar)" name="ydlidar_lidar_publisher"  pkg="ydlidar_ros_driver"  type="ydlidar_ros_driver_node" output="log" respawn="false" >
    <!-- string property -->
    <param name="port"         type="string" value="/dev/ydlidar"/>  
    <param name="frame_id"     type="string" value="laser_frame"/>
    <param name="ignore_array"     type="string" value=""/>

    <!-- int property -->
    <param name="baudrate"         type="int" value="128000"/>  
    <!-- 0:TYPE_TOF, 1:TYPE_TRIANGLE, 2:TYPE_TOF_NET -->
    <param name="lidar_type"       type="int" value="1"/>  
    <!-- 0:YDLIDAR_TYPE_SERIAL, 1:YDLIDAR_TYPE_TCP -->
    <param name="device_type"         type="int" value="0"/>  
    <param name="sample_rate"         type="int" value="5"/>  
    <param name="abnormal_check_count"         type="int" value="4"/>  

    <!-- bool property -->
    <param name="resolution_fixed"    type="bool"   value="true"/>
    <param name="auto_reconnect"    type="bool"   value="true"/>
    <param name="reversion"    type="bool"   value="false"/>
    <param name="inverted"    type="bool"   value="true"/>
    <param name="isSingleChannel"    type="bool"   value="false"/>
    <param name="intensity"    type="bool"   value="false"/>
    <param name="support_motor_dtr"    type="bool"   value="true"/>
    <param name="invalid_range_is_inf"    type="bool"   value="false"/>
    <param name="point_cloud_preservative"    type="bool"   value="false"/>

    <!-- float property -->
    <param name="angle_min"    type="double" value="-180" />
    <param name="angle_max"    type="double" value="180" />
    <param name="range_min"    type="double" value="0.1" />
    <param name="range_max"    type="double" value="12.0" />
    <!-- frequency is invalid, External PWM control speed -->
    <param name="frequency"    type="double" value="10.0"/>
  </node>

  <!-- <node pkg="tf" type="static_transform_publisher" name="base_link_to_laser4"
    args="0.0 0.0 0.2 0.0 0.0 0.0 /base_footprint /laser_frame 40" /> -->
</launch>